crates/file_finder/src/file_finder.rs

```diff
@@ -230,88 +274,104 @@ impl Matches {
+            .map(Match::Search)
+            .collect();
 
-                    (Match::History(_, _), Match::Search(_)) if self.separate_history => {
-                        cmp::Ordering::Less
-                    }
-                    (Match::Search(_), Match::History(_, _)) if self.separate_history => {
-                        cmp::Ordering::Greater
-                    }
+        if extend_old_matches {
+            //NOTE: since we take history matches instead of new search matches
+            // and history matches has not changed(since the query has not changed and we do not extend old matches otherwise),
+            // old matches can't contain paths present in history_matches as well.
+            self.matches.retain(|m| matches!(m, Match::Search(_)));
+        } else {
+            self.matches.clear();
+        }
 
-                    (Match::History(_, match_a), Match::History(_, match_b)) => {
-                        match_b.cmp(match_a)
-                    }
-                    (Match::History(_, match_a), Match::Search(match_b)) => {
-                        Some(match_b).cmp(&match_a.as_ref())
-                    }
-                    (Match::Search(match_a), Match::History(_, match_b)) => {
-                        match_b.as_ref().cmp(&Some(match_a))
+        //NOTE: At this point we have an unsorted set of new history matches, an unsorted set of new search matches
+        // and a sorted set of old search matches.
+        // It is possible that the new search matches' paths contain some of the old search matches' paths.
+        // History matches' paths are unique, since store in a HashMap by path.
+        // We build a sorted Vec<Match>, eliminating duplicate search matches.
+        // Search matches with the same paths should have equal `ProjectPanelOrdMatch`, so we should
+        // not have any duplicates after building the final list.
+        for new_match in new_history_matches
+            .into_values()
+            .chain(new_search_matches.into_iter())
+        {
+            match self.position(&new_match) {
+                Ok(_duplicate) => continue,
+                Err(i) => {
+                    self.matches.insert(i, new_match);
+                    if self.matches.len() == 100 {
+                        break;
                     }
-                    (Match::Search(match_a), Match::Search(match_b)) => match_b.cmp(match_a),
                 }
-                .then(history_score_a.cmp(history_score_b))
-            })
-            .take(100)
-            .map(|(_, m)| m)
-            .collect();
+            }
+        }
+    }
+
+    /// If a < b, then a is a worse match, aligning with the `ProjectPanelOrdMatch` ordering.
+    fn cmp_matches(
+        separate_history: bool,
+        currently_opened: Option<&FoundPath>,
+        a: &Match,
+        b: &Match,
+    ) -> cmp::Ordering {
+        assert!(a.panel_match().is_some() && b.panel_match().is_some());
```

Comment:
```suggestion
        debug_assert!(a.panel_match().is_some() && b.panel_match().is_some());
```

I'm a bit weary of panics in production and trying not to introduce more potential ones.